### PR TITLE
Add support for Red Hat Bugzilla

### LINKED ISSUE #4: RFE: Please add support for Red Hat Bugzilla
Please consider adding support for [Red Hat Bugzilla](https://bugzilla.redhat.com/) - [API Docs](https://bugzilla.readthedocs.io/en/latest/api/core/v1/general.html#).

Differences that I'm aware of are in Bug object:

* `component` property is of type `string[]` - _currently only `string` is supported_ https://github.com/Mossop/bugzilla-ts/blob/36fcd9a4b825c9a78b4f0ceba5ab34d6f430c81c/src/types.ts#L60
* `flags` property could be `undefined` - _currently validators expect only empty array_ https://github.com/Mossop/bugzilla-ts/blob/36fcd9a4b825c9a78b4f0ceba5ab34d6f430c81c/src/types.ts#L66
* `version` property is of type `string[]` - _currently only `string` is supported_ https://github.com/Mossop/bugzilla-ts/blob/36fcd9a4b825c9a78b4f0ceba5ab34d6f430c81c/src/types.ts#L89

---

**How to reproduce:**

```ts
import BugzillaAPI from 'bugzilla';

let api = new BugzillaAPI('https://bugzilla.redhat.com/');

api
  .getBugs([1556154, 1906064])
  .then(bug => console.log(bug))
  .catch(err => console.log(err));
```

---

**Example Request `Fedora` product:**

```url
GET https://bugzilla.redhat.com/rest/bug/123456
```

<details>
<summary>Response</summary>
<br>
<pre>
{
   "bugs" : [
      {
         "alias" : [],
         "assigned_to" : "Arjan van de Ven",
         "assigned_to_detail" : {
            "email" : "arjanv",
            "id" : 29238,
            "name" : "arjanv",
            "real_name" : "Arjan van de Ven"
         },
         "blocks" : [],
         "cc" : [
            "Blade_Arma",
            "Issue Tracker"
         ],
         "cc_detail" : [
            {
               "email" : "edgbla",
               "id" : 315394,
               "name" : "edgbla",
               "real_name" : "Blade_Arma"
            },
            {
               "email" : "tao",
               "id" : 80926,
               "name" : "tao",
               "real_name" : "Issue Tracker"
            }
         ],
         "cf_clone_of" : null,
         "cf_doc_type" : "Bug Fix",
         "cf_environment" : "",
         "cf_fixed_in" : "",
         "cf_last_closed" : "2006-02-21T19:03:16Z",
         "cf_release_notes" : "",
         "cf_type" : "---",
         "classification" : "Fedora",
         "component" : [
            "kernel"
         ],
         "creation_time" : "2004-05-18T19:39:17Z",
         "creator" : "Gil Chilton",
         "creator_detail" : {
            "email" : "gil.chilton",
            "id" : 30769,
            "name" : "gil.chilton",
            "real_name" : "Gil Chilton"
         },
         "deadline" : null,
         "depends_on" : [],
         "docs_contact" : "",
         "dupe_of" : 123935,
         "groups" : [],
         "id" : 123456,
         "is_cc_accessible" : true,
         "is_confirmed" : true,
         "is_creator_accessible" : true,
         "is_open" : false,
         "keywords" : [],
         "last_change_time" : "2016-07-05T15:12:23Z",
         "op_sys" : "Linux",
         "platform" : "i386",
         "priority" : "medium",
         "product" : "Fedora",
         "qa_contact" : "",
         "resolution" : "DUPLICATE",
         "see_also" : [],
         "severity" : "medium",
         "status" : "CLOSED",
         "summary" : "System with Syntax - S635MP motherboard will not install",
         "target_milestone" : "---",
         "target_release" : [
            "---"
         ],
         "url" : "",
         "version" : [
            "2"
         ],
         "whiteboard" : ""
      }
   ],
   "faults" : []
}
</pre>
</details>

---

**Example Request `RHEL` product:**

```url
GET https://bugzilla.redhat.com/rest/bug/1906064
```

**Response:**

<details>
<summary>Response</summary>
<br>
<pre>
{
   "bugs" : [
      {
         "alias" : [],
         "assigned_to" : "Jan Macku",
         "assigned_to_detail" : {
            "email" : "jamacku",
            "id" : 431924,
            "name" : "jamacku",
            "real_name" : "Jan Macku"
         },
         "blocks" : [
            1997981,
            1997986
         ],
         "cc" : [
            "Jan Macku"
         ],
         "cc_detail" : [
            {
               "email" : "jamacku",
               "id" : 431924,
               "name" : "jamacku",
               "real_name" : "Jan Macku"
            }
         ],
         "cf_clone_of" : null,
         "cf_doc_type" : "If docs needed, set a value",
         "cf_environment" : "",
         "cf_fixed_in" : "",
         "cf_last_closed" : null,
         "cf_release_notes" : "",
         "cf_target_upstream_version" : "",
         "classification" : "Red Hat",
         "component" : [
            "tcsh"
         ],
         "creation_time" : "2020-12-09T15:51:23Z",
         "creator" : "Frank Hirtz",
         "creator_detail" : {
            "email" : "fhirtz",
            "id" : 30797,
            "name" : "fhirtz",
            "real_name" : "Frank Hirtz"
         },
         "deadline" : null,
         "depends_on" : [],
         "docs_contact" : "",
         "dupe_of" : null,
         "groups" : [],
         "id" : 1906064,
         "is_cc_accessible" : true,
         "is_confirmed" : true,
         "is_creator_accessible" : true,
         "is_open" : true,
         "keywords" : [],
         "last_change_time" : "2021-12-01T15:18:46Z",
         "op_sys" : "All",
         "platform" : "All",
         "priority" : "high",
         "product" : "Red Hat Enterprise Linux 7",
         "qa_contact" : "Karel Volný",
         "qa_contact_detail" : {
            "email" : "kvolny",
            "id" : 213715,
            "name" : "kvolny",
            "real_name" : "Karel Volný"
         },
         "resolution" : "",
         "see_also" : [],
         "severity" : "high",
         "status" : "POST",
         "summary" : "tcsh history breaks when using modifiers",
         "target_milestone" : "rc",
         "target_release" : [
            "---"
         ],
         "url" : "",
         "version" : [
            "7.9"
         ],
         "whiteboard" : ""
      }
   ],
   "faults" : []
}
</pre>
</details>

---

Thank you for your work!

### PR DESCRIPTION
**Pull request by jamacku**
Fixes #4

Thank you!

### CODE REVIEW

File: src/validators.ts [hunk 1]
```diff
@@ -114,3 +114,21 @@ export function optional<T>(
     return validator(val);
   };
 }
+
+export function optionalArray<T>(validator: Validator<T>): Validator<T | T[]>;
```
Comment (Mossop) [inline src/validators.ts]: Elsewhere `optional` means `T | undefined` so at a glance I would expect `optionalArray(...)` to be the same as `optional(array(...))`. So I'd prefer to use a different term here. How about `maybeArray`?

File: src/validators.ts [hunk 2]
```diff
@@ -114,3 +114,21 @@ export function optional<T>(
...
+  validator: Validator<T>,
+  result: T | T[],
+): Validator<T | T[]>;
+export function optionalArray<T>(
+  validator: Validator<T>,
+  result?: T | T[],
+): Validator<T | T[] | undefined> {
+  return (val: any): T | T[] => {
+    if (Array.isArray(val)) {
+      return result ?? val;
```
Comment (Mossop) [inline src/validators.ts]: I'm a little confused by this, if the received value is an array this defaults to returning a passed default result?

I think we just don't need the default result at all and instead want to run the validator over every element in the array if `val` is an array.

**Review by Mossop — CHANGES_REQUESTED** [review]
It's unfortunate that different bugzilla installs can have different field types like this but I've not thought of a better solution to that so this seems like a reasonable change.

File: src/validators.ts (hunk 1 above)
Comment (jamacku) [reply src/validators.ts]: That's reasonable, I  update name accordingly, Thanks

File: src/validators.ts (hunk 2 above)
Comment (jamacku) [reply src/validators.ts]: I'm sorry for confusion. I'm new to Typescript generics and those complex types, I'll try to update patch.

**Comment by jamacku** [comment]
@Mossop I updated validator, could you please have a look? Thanks!

**Comment by Mossop** [comment]
Looks good, thanks!

**Pull request merged**